service/gcs/bridge/bridge_handler_test.go

```diff
@@ -0,0 +1,791 @@
+	logrus.SetOutput(ioutil.Discard)
+
+	r, hs := createContainerConfig()
+	req, rw := setupRequestResponse(t, prot.ComputeSystemCreateV1, r)
+
+	mc := &mockcore.MockCore{Behavior: mockcore.SingleSuccess}
+	mc.WaitContainerWg.Add(1)
+
+	tb := &Bridge{coreint: mc}
+	tb.createContainer(rw, req)
+
+	verifyResponseSuccess(t, rw)
+	if r.ContainerID != mc.LastCreateContainer.ID {
+		t.Fatal("last create container did not have the same container ID")
+	}
+	if !reflect.DeepEqual(hs, mc.LastCreateContainer.Settings) {
+		t.Fatal("request/response structs are not equal")
+	}
+
+	// Verify that wait was called. This also tests that if we dont exit in the
+	// error case here we would panic when PublishNotification tries to write to
+	// the responseChan.
+	mc.WaitContainerWg.Wait()
+}
+
+func Test_CreateContainer_Success_WaitContainer_Success(t *testing.T) {
+	r, hs := createContainerConfig()
+	req, rw := setupRequestResponse(t, prot.ComputeSystemCreateV1, r)
+
+	mc := &mockcore.MockCore{Behavior: mockcore.Success}
+	mc.WaitContainerWg.Add(1)
+	b := &Bridge{coreint: mc}
+	b.responseChan = make(chan bridgeResponse)
+	defer close(b.responseChan)
+
+	publishWg := sync.WaitGroup{}
+	publishWg.Add(1)
+	go func() {
+		defer publishWg.Done()
+
+		response := <-b.responseChan
+
+		cn := response.response.(*prot.ContainerNotification)
+		if cn.ContainerID != r.ContainerID {
+			t.Fatal("publish response had invalid container ID")
+		}
+		if cn.ActivityID != r.ActivityID {
+			t.Fatal("publish response had invalid activity ID")
+		}
+		if cn.Type != prot.NtUnexpectedExit {
+			t.Fatal("publish response had invalid type")
+		}
+		if cn.Operation != prot.AoNone {
+			t.Fatal("publish response had invalid operation")
+		}
+		if cn.Result != -1 {
+			t.Fatal("publish response had invalid result")
+		}
+	}()
+
+	b.createContainer(rw, req)
+	verifyResponseSuccess(t, rw)
+	if r.ContainerID != mc.LastCreateContainer.ID {
+		t.Fatal("last create container did not have the same container ID")
+	}
+	if !reflect.DeepEqual(hs, mc.LastCreateContainer.Settings) {
+		t.Fatal("last create container did not have equal settings structs")
+	}
+
+	mc.WaitContainerWg.Wait()
+	if r.ContainerID != mc.LastWaitContainer.ID {
+		t.Fatal("last wait container did not have the same container ID")
+	}
+
+	// Wait for the publish to take place on the exited notification.
+	publishWg.Wait()
+}
+
+func Test_ExecProcess_InvalidJson_Failure(t *testing.T) {
+	req, rw := setupRequestResponse(t, prot.ComputeSystemExecuteProcessV1, nil)
+
+	tb := new(Bridge)
+	tb.execProcess(rw, req)
+
+	verifyResponseJSONError(t, rw)
+}
+
+func Test_ExecProcess_InvalidProcessParameters_Failure(t *testing.T) {
+	r := &prot.ContainerExecuteProcess{
+		MessageBase: newMessageBase(),
+		Settings: prot.ExecuteProcessSettings{
+			ProcessParameters: "",
+		},
+	}
+
+	req, rw := setupRequestResponse(t, prot.ComputeSystemExecuteProcessV1, r)
+
+	tb := new(Bridge)
+	tb.execProcess(rw, req)
+
+	verifyResponseJSONError(t, rw)
```

Comment:
You should probably test for correct ActivityId on both success and failure as well. I think not filling in ActivityId for failures is the cause on the invalid JSON error we've been seeing.

Reply:
Fixed in latest push.